source/distributed/grid_tools.cc

```diff
@@ -0,0 +1,166 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/point.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_tools.h>
+
+#ifdef DEAL_II_WITH_P4EST
+
+#include <vector>
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace GridTools
+  {
+    template < int dim, int spacedim>
+    std::vector<BoundingBox<spacedim> >
+    compute_locally_owned_bounding_box
+    (const parallel::distributed::Triangulation< dim, spacedim > &distributed_tria)
+    {
+      // In order to return the boundary points of the locally_owned cells correctly
+      // we need to handle the possibility of a second connected component.
+      //
+      // Assuming the triangulation is distributed following the morton-z curve order
+      // we know that, if there are two connected components, then the first and the
```

Comment:
please assert that all locally owned cells are exactly within [firstc, lastc] in debug mode. I am not sure this is correct, by the way. We do not iterate over cells in morton ordering.